File: libView/src/main/java/org/waynezhou/libView/recycler/RecyclerList.java
package org.waynezhou.libView.recycler;

import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GestureDetectorCompat;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import org.waynezhou.libUtil.LogHelper;
import org.waynezhou.libUtil.reflection.ReflectionException;
import org.waynezhou.libView.LibView;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.ListIterator;

public class RecyclerList<TItem, TViewHolder extends RecyclerView.ViewHolder>
    extends RecyclerView.Adapter<TViewHolder>
    implements List<TItem>
{
    private final AppCompatActivity activity;
    private final RecyclerListBinder<TItem, TViewHolder> binder;
    private final List<TItem> source;
    public RecyclerList(AppCompatActivity activity, RecyclerListBinder<TItem, TViewHolder> binder, List<TItem> keeper){
        this.activity = activity;
        this.binder = binder;
        this.source = new LinkedList<>(keeper);
    }
    public RecyclerList(AppCompatActivity activity, RecyclerListBinder<TItem, TViewHolder> binder){
        this(activity, binder, new ArrayList<>());
    }
    
    @NonNull
    @Override
    public TViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return binder.createViewHolder(parent);
    }
    
    @Override
    public void onBindViewHolder(@NonNull TViewHolder holder, int position) {
        binder.onBind(holder, source, position);
    }
    
    @Override
    public int getItemCount() {
        return size();
    }
    
    
    // region list
    // region region readonly methods of implement [List<TItem>]
    @Override
    public int size() {
        return source.size();
    }
    
    @Override
    public boolean isEmpty() {
        return source.isEmpty();
    }
    
    @Override
    public boolean contains(@Nullable Object o) {
        return source.contains(o);
    }
    
    @NonNull
    @Override
    public Iterator<TItem> iterator() {
        return source.iterator();
    }
    
    @NonNull
    @Override
    public Object[] toArray() {
        return source.toArray();
    }
    
    @SuppressWarnings("SuspiciousToArrayCall")
    @NonNull
    @Override
    public <T> T[] toArray(@NonNull T[] a) {
        return source.toArray(a);
    }
    
    @Override
    public boolean containsAll(@NonNull Collection<?> c) {
        return source.containsAll(c);
    }
    
    @Override
    public TItem get(int index) {
        return source.get(index);
    }
    
    @Override
    public int indexOf(@Nullable Object o) {
        return source.indexOf(o);
    }
    
    @Override
    public int lastIndexOf(@Nullable Object o) {
        return source.lastIndexOf(o);
    }
    
    @NonNull
    @Override
    public ListIterator<TItem> listIterator() {
        return source.listIterator();
    }
    
    @NonNull
    @Override
    public ListIterator<TItem> listIterator(int index) {
        return source.listIterator(index);
    }
    
    @NonNull
    @Override
    public List<TItem> subList(int fromIndex, int toIndex) {
        return source.subList(fromIndex, toIndex);
    }
    // endregion
    
    // region mutable methods of implement [MutableList<TItem>]
    // with injected notify index
    @Override
    public boolean add(TItem tItem) {
        boolean result = source.add(tItem);
        activity.runOnUiThread(() -> notifyItemInserted(source.size() - 1));
        return result;
    }
    
    @SuppressWarnings("SuspiciousMethodCalls")
    @Override
    public boolean remove(@Nullable Object o) {
        int removedIdx = source.indexOf(o);
        if(removedIdx==-1) return false;
        boolean result = source.remove(o);
        activity.runOnUiThread(() -> notifyItemRemoved(removedIdx));
        return result;
    }
    
    @Override
    public boolean addAll(@NonNull Collection<? extends TItem> c) {
        int from = source.size();
        boolean result = source.addAll(c);
        activity.runOnUiThread(() -> notifyItemRangeInserted(from, c.size() ));
        return result;
    }
    
    @Override
    public boolean addAll(int index, @NonNull Collection<? extends TItem> c) {
        boolean result = source.addAll(index, c);
        activity.runOnUiThread(() -> notifyItemRangeInserted(index,  c.size() ));
        return result;
    }
    
    @Override
    public boolean removeAll(@NonNull Collection<?> c) {
        boolean result = source.removeAll(c);
        activity.runOnUiThread(this::notifyDataSetChanged);
        return result;
    }
    
    @Override
    public boolean retainAll(@NonNull Collection<?> c) {
        boolean result = source.retainAll(c);
        activity.runOnUiThread(this::notifyDataSetChanged);
        return result;
    }
    
    @Override
    public void clear() {
        source.clear();
        activity.runOnUiThread(this::notifyDataSetChanged);
    }
    
    @Override
    public TItem set(int index, TItem element) {
        TItem result = source.set(index, element);
        activity.runOnUiThread(() -> notifyItemChanged(index));
        return result;
    }
    
    @Override
    public void add(int index, TItem element) {
        source.add(index, element);
        activity.runOnUiThread(() -> notifyItemInserted(index));
    }
    
    @Override
    public TItem remove(int index) {
        TItem result = source.remove(index);
        activity.runOnUiThread(() -> notifyItemRemoved(index));
        return result;
    }
    // endregion
    // endregion
}

File: libViewKt/src/main/java/org/waynezhou/libviewkt/RecyclerList.kt
package org.waynezhou.libviewkt


import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.SurfaceHolder
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import org.waynezhou.libutilkt.LogHelper
import org.waynezhou.libutilkt.reflection.ReflectionException
import org.waynezhou.libviewkt.LibView.dataContext

class RecyclerList<TItem, TViewHolder: RecyclerView.ViewHolder>
constructor(
    private val activity: AppCompatActivity,
    private val binder: RecyclerListBinder<TItem, TViewHolder>,
    keeper: List<TItem> = listOf(),
) : RecyclerView.Adapter<TViewHolder>(), MutableList<TItem> {
    private val source = keeper.toMutableList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TViewHolder {
        return binder.createViewHolder(parent)
    }

    override fun onBindViewHolder(holder: TViewHolder, position: Int) {
        binder.onBind(holder, source, position);
    }

    override fun getItemCount(): Int {
        return size;
    }

    //region list readonly implement
    override val size: Int
        get() = source.size

    override fun contains(element: TItem): Boolean = source.contains(element)

    override fun containsAll(elements: Collection<TItem>): Boolean = source.containsAll(elements)

    override fun get(index: Int): TItem = source[index]

    override fun indexOf(element: TItem): Int  = source.indexOf(element)

    override fun isEmpty(): Boolean = source.isEmpty()

    override fun iterator(): MutableIterator<TItem> = source.iterator()

    override fun lastIndexOf(element: TItem): Int = source.lastIndexOf(element)


    override fun listIterator(): MutableListIterator<TItem> = source.listIterator()

    override fun listIterator(index: Int): MutableListIterator<TItem> = source.listIterator(index)

    override fun subList(fromIndex: Int, toIndex: Int): MutableList<TItem>  = source.subList(fromIndex, toIndex)

    //endregion

    // region list mutable implement with notify
    override fun add(element: TItem): Boolean = source.add(element).apply {
        activity.runOnUiThread{notifyItemChanged(size-1)}
    }

    override fun add(index: Int, element: TItem) = source.add(index, element).apply {
        activity.runOnUiThread{notifyItemChanged(index)}
    }

    override fun addAll(index: Int, elements: Collection<TItem>): Boolean = source.addAll(index, elements).apply {
        activity.runOnUiThread{notifyItemRangeChanged(index, elements.size)}
    }

    override fun addAll(elements: Collection<TItem>): Boolean = source.addAll(elements).apply {
        activity.runOnUiThread{notifyItemRangeChanged(size-elements.size, elements.size)}
    }

    override fun clear() {
        val count = size
        source.clear()
        activity.runOnUiThread{notifyItemRangeRemoved(0, count)}
    }

    override fun remove(element: TItem): Boolean {
        val idx = source.indexOf(element)
        val ret = source.remove(element)
        activity.runOnUiThread{notifyItemRemoved(idx)}
        return ret
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun removeAll(elements: Collection<TItem>): Boolean = source.removeAll(elements).run {
        activity.runOnUiThread { notifyDataSetChanged() }
        this
    }

    override fun removeAt(index: Int): TItem = source.removeAt(index).run {
        activity.runOnUiThread { notifyItemRemoved(index) }
        this
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun retainAll(elements: Collection<TItem>): Boolean= source.removeAll(elements).run {
        activity.runOnUiThread { notifyDataSetChanged() }
        this
    }

    override fun set(index: Int, element: TItem): TItem = source.set(index, element).run {
        activity.runOnUiThread { notifyItemChanged(index) }
        this
    }
    // endregion
}

abstract class RecyclerListBinder<TItem, TViewHolder: RecyclerView.ViewHolder>{


    abstract fun createViewHolder(parent: ViewGroup): TViewHolder
    abstract fun onBind(holder: TViewHolder, items:List<TItem>, position: Int)
}

open class ViewBindingRecyclerListBinder<TItem, TViewBinding: ViewBinding>
    (private val inflater: LayoutInflater, private val viewBindingClass: Class<TViewBinding>, private val binder: (TViewBinding, List<TItem>, position:Int)->Unit)
    : RecyclerListBinder<TItem, ViewBindingRecyclerListBinder<TItem, TViewBinding>.ViewHolder>(){

    inner class ViewHolder(var binding: TViewBinding): RecyclerView.ViewHolder(binding.root)

    private val holderMap: MutableMap<Int, ViewHolder> = mutableMapOf()

    override fun createViewHolder(parent: ViewGroup): ViewHolder {
        return ViewHolder(LibView.inflate(inflater, viewBindingClass, parent, false))
    }

    override fun onBind(holder:ViewHolder, items: List<TItem>, position: Int) {
        binder(holder.binding, items, position)
    }
}

class ViewDataBindingRecyclerListBinder<TItem, TViewBinding: ViewDataBinding>
    ( inflater: LayoutInflater, viewBindingClass: Class<TViewBinding>, private val binder: TViewBinding.(holder:RecyclerView.ViewHolder, item:TItem)->Unit)
    : ViewBindingRecyclerListBinder<TItem, TViewBinding>(inflater, viewBindingClass, { _,_,_ -> }) {

    override fun createViewHolder(parent: ViewGroup): ViewHolder {
        val holder = super.createViewHolder(parent)
        holder.binding = DataBindingUtil.bind(holder.itemView)!!
        return holder
    }

    override fun onBind(holder:ViewHolder, items: List<TItem>, position: Int) {
        LogHelper.d(position)
        binder(holder.binding, holder, items[position])
    }
}
